### ISSUE TITLE
Xml_generator: prevent unnecessary new lines

### ISSUE THREAD

**Issue by m-stein** [genodelabs/genode#2967]
The handling of exceptions in the Xml_generator::Node constructor leaves unwanted new lines in the XML string.

**Comment by nfeske** [comment]
This issue title, description, and label leaves me puzzled. It suggests that something fundamental is broken and must be fixed but remains completely unspecific about the symptom and nature of the problem.

If the issue is about the topic we discussed offline just yesterday, an appropriate title would be "Xml_generator: improve formatting on exception" with the label "cleanup" attached. From what I know, the issue is merely about a formatting improvement in the exception case, which happened to become a relevant in your case where exceptions are used in a non-exceptional way.

I want to encourage you to spend a little more effort on meaningful issue titles, descriptions, and labels. An issue should at least be specific about the symptom and the expected behavior. Please keep in mind that the issue tracker is part of our project's documentation.

**Comment by m-stein** [comment]
@nfeske Sorry for that. I thought it would be fixed right away but it turned out to be more time-intensive.

**Comment by nfeske** [comment]
The commits are very insightful. Thanks for the good commit messages and the extended tests.

**Comment by chelmuth** [comment]
I suspect the commit series breaks `make run/report_rom`, please check.

**Comment by nfeske** [comment]
The commit `Xml_generator: fix and test missing '\0'` causes a segfault in init in the leitzentrale.run scenario on base-linux.

**Comment by nfeske** [comment]
Let's keep this line of work for merging at a later point because it deserves more attention than I'm able to provide right now.

@m-stein I'm wondering, in the commit message you state "Without exceptions in the Node constructor this led to "\0\n" at the end of generated XML." as a fact. But have you actually witnessed this sequence in the wild? I'm asking because I haven't (looking at the reports in sculpt's report fs), and when looking at the code, I would not expect it either.

Let my try to clarify, the `_out_buffer.append('\')` operation is applied to a local copy of the `_out_buffer` _after_ committing the content. This local copy (the buffer meta data, not the characters) is discarded afterwards. Hence, the zero-character is inserted but it stays outside the committed character sequence.

BTW, The `\n` at the end ensures that the generated data ends with a newline when printed, e.g., by `cat /report/runtime/state`. I would like to preserve it. (I noticed that you added a manual `\n` to the test, which I find odd). But it actually overwrites the last written `\0`. So it might be sensible to also append a `\0` here. That said, we normally don't depend on the zero-termination of the generated string but rather propagate the number of generated characters to the user via the `user()` accessor (see the use in repos/os/include/os/reporter.h). So I'm still puzzled how your patch could introduce the segfault problem I reported for `leitzentrale.run`.

...to be investigated later.
